src/fsharp/infos.fs

```diff
@@ -122,58 +119,95 @@ let rec GetImmediateInterfacesOfType skipUnref g amap m ty =
-                                   yield ity
-                  | _ -> ()
-                  yield mkAppTy g.system_GenericIComparable_tcref [ty]
-                  yield mkAppTy g.system_GenericIEquatable_tcref [ty]]
-            else
-                getInterfaces ty tcref tinst
-        | _ ->
+            // Check if this is a measure-annotated type
+            match tcref.TypeReprInfo with
+            | TMeasureableRepr reprTy ->
+                yield! GetImmediateInterfacesOfMeasureAnnotatedType skipUnref g amap m ty reprTy
+            | _ ->
+                yield! GetImmediateInterfacesOfMetadataType g amap m skipUnref ty tcref tinst
+
+        | ValueNone ->
+            // For tuple types, func types, check if we can eliminate to a type with metadata.
             let tyWithMetadata = convertToTypeWithMetadataIfPossible g ty
             match tryAppTy g tyWithMetadata with
             | ValueSome (tcref, tinst) ->
                 if isAnyTupleTy g ty then
-                    getInterfaces tyWithMetadata tcref tinst
-                else
-                    []
-            | _ -> []
-
-    // NOTE: Anonymous record types are not directly considered to implement IComparable,
-    // IComparable<T> or IEquatable<T>. This is because whether they support these interfaces depend on their
-    // consitutent types, which may not yet be known in type inference.
-    //
-    // NOTE: Tuples could in theory always support IComparable etc. because this
-    // is in the .NET metadata for System.Tuple etc.  However from the F# perspective tuple types don't
-    // always support the 'comparable' and 'equality' constraints (again, it depends on their constitutent types).
+                    yield! GetImmediateInterfacesOfMetadataType g amap m skipUnref tyWithMetadata tcref tinst
+            | _ -> ()
 
-    // .NET array types are considered to implement IList<T>
-    let itys =
+        // .NET array types are considered to implement IList<T>
         if isArray1DTy g ty then
-            mkSystemCollectionsGenericIListTy g (destArrayTy g ty) :: itys
-        else
-            itys
-
-    itys
+            mkSystemCollectionsGenericIListTy g (destArrayTy g ty)
+    ]
+
+// Report the interfaces supported by a measure-annotated type.
+//
+// For example, consider:
+//
+//     [<MeasureAnnotatedAbbreviation>]
+//     type A<[<Measure>] 'm> = A
+//
+// This measure-annotated type is considered to support the interfaces on its representation type A,
+// with the exception that
+//
+//   1. we rewrite the IComparable and IEquatable interfaces, so that
+//    IComparable<A> --> IComparable<A<'m>>
+//    IEquatable<A> --> IEquatable<A<'m>>
+//
+//   2. we emit any other interfaces that derive from IComparable and IEquatable interfaces
+//
+// This rule is conservative and only applies to IComparable and IEquatable interfaces.
+//
+// This rule may in future be extended to rewrite the "trait" interfaces associated with .NET 7.
```

Comment:
Yeah, we'll probably want to revisit what are we looing for, once it's stabilized in .NET7+